spec/services/action_create_service_spec.rb

```diff
@@ -60,9 +61,23 @@
       stage2.reload
       request.reload
       expect(action1).to have_attributes(:operation => Action::DENY_OPERATION, :processed_by => 'man', :comments => 'bad')
-      expect(stage1).to  have_attributes(:state => Stage::FINISHED_STATE, :decision => Stage::DENIED_STATUS, :reason => 'bad')
-      expect(stage2).to  have_attributes(:state => Stage::SKIPPED_STATE,  :decision => Stage::UNDECIDED_STATUS, :reason => nil)
-      expect(request).to have_attributes(:state => Stage::FINISHED_STATE, :decision => Stage::DENIED_STATUS, :reason => 'bad')
+      expect(stage1).to  have_attributes(:state => Stage::FINISHED_STATE,   :decision => Stage::DENIED_STATUS, :reason => 'bad')
+      expect(stage2).to  have_attributes(:state => Stage::SKIPPED_STATE,    :decision => Stage::UNDECIDED_STATUS, :reason => nil)
+      expect(request).to have_attributes(:state => Request::FINISHED_STATE, :decision => Request::DENIED_STATUS, :reason => 'bad')
+    end
+  end
+
+  context 'cancel operation' do
+    it 'updates stage and request' do
+      stage1.update_attributes(:state => Stage::NOTIFIED_STATE)
+      action1 = svc1.create('operation' => Action::CANCEL_OPERATION, 'processed_by' => 'requester', 'comments' => 'regret')
+      stage1.reload
+      stage2.reload
+      request.reload
+      expect(action1).to have_attributes(:operation => Action::CANCEL_OPERATION, :processed_by => 'requester', :comments => 'regret')
+      expect(stage1).to  have_attributes(:state => Stage::CANCELED_STATE,   :decision => Stage::UNDECIDED_STATUS, :reason => 'regret')
+      expect(stage2).to  have_attributes(:state => Stage::SKIPPED_STATE,    :decision => Stage::UNDECIDED_STATUS, :reason => nil)
```

Comment:
I'm wondering if `stage2` should also be set to `CANCELED_STATE` as opposed to `SKIPPED_STATE`? Same with the `decision` as `UNDECIDED_STATUS` for each.  It seems like all the stages should have the cancelled state and the reason set to a 'CANCELED_STATUS'.  

In my mind though, I'm seeing a 'cancel` as a definite action with a reason of 'canceled'.  Possibly that is not how the backend behavior models this decision.

Reply:
The backend implementation has a distinguish of system role vs human role. `skip` action and its corresponding `skipped` state are done by system (PAM in this case), while other actions such as `approve`, `deny`, and `cancel` are done by human. If a certain stage is `denied` by human then the system decides to `skip` the rest stages. Here we are following the same pattern for canceling. No code change is needed, i.e., the system always skips unneeded stages, not cancel.

As for the status, I agree we may be able to introduce a `CANCELED_STATUS`.